include/aws/common/byte_buf.h

```diff
@@ -39,6 +40,34 @@ static inline struct aws_byte_buf aws_byte_buf_from_array(const uint8_t c_str[],
     return (struct aws_byte_buf){.buffer = (uint8_t *)&c_str[0], .len = len};
 }
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * No copies, no string allocations. Fills in output with a list of aws_byte_buf instances where buffer is
```

Comment:
That's not my intention with byte_buffer. It was merely to join len and array into a single object and to not pass c strings around.

Reply:
The way the aws_byte_cursor_advance* functions are written, they are changing the value of the pointer and updating the length to be the length remaining in the array. @bdonlan was originally using the byte_buffer struct to keep that pointer and length together as a single entity, but that concerned me, because if we're dealing with heap-allocated memory we would want to keep a pointer to the beginning of it for freeing later. Also, I felt that the name "buffer" implied that this was the entire array of memory and not just something pointing to somewhere in the middle of it. I could easily see someone who picked this code up later getting confused and trying to free the memory somewhere in the middle of the array and causing a crash.

This is why we changed those functions to be working with the byte cursor, where the struct for the cursor is essentially the same (a pointer and a length), but the pointer might point to anywhere in the middle of the array, and so that we knew never to free the memory through the cursor. Whereas the plan was to keep the byte buffer objects always pointing to the beginning of the allocated memory with the length field representing their full length.

I understand these concerns and decisions were not what motivated the initial creation of the byte_buffer struct, but those design decisions have already made their way into the byte_buf.h code in this package, specifically in those byte cursor advance functions.

I can see it either way here, the substrings shouldn't be freed themselves, but you might want to keep them as byte buffers and then only create byte cursor objects when you were reading through them. (We've been in the habit of *always* reading the data in our byte buffers with the byte cursor advance functions to ensure that we are not reading off the edges.)  Or you might want to call them byte cursors from the beginning to differentiate that they shouldn't be freed.

Mostly, I just want to make sure that we agree on the conventions so that we don't end up working at cross-purposes on the same code base.

Reply:
Let me add that I won't hold up the pull request on this issue. Just want to make sure you think about it.